### PR TITLE
Use epoll in Netty if epoll is available.

### PR DESCRIPTION
**Pull request by zzmao**
Epoll should be use in platforms where Epoll are supported.

### CODE REVIEW

**Comment by zzmao** [comment]
Yes, we need. 

BTW, do you know why we didn't epoll this before? In Frontend NettyServer.

**Comment by cgtz** [comment]
I'm not too sure, but I see that FrontendIntegrationTest got an exception on travis. @vgkholla, do you know of any known incompatibilities with EpollEventLoopGroup?

File: ambry-rest/src/main/java/com/github/ambry/rest/NettyServer.java [hunk 1]
```diff
@@ -72,8 +74,14 @@ public void start() throws InstantiationException {
...
-      bossGroup = new NioEventLoopGroup(nettyConfig.nettyServerBossThreadCount);
-      workerGroup = new NioEventLoopGroup(nettyConfig.nettyServerWorkerThreadCount);
+      if (Epoll.isAvailable()) {
+        logger.trace("Using EpollEventLoopGroup in NettyServer.");
+        bossGroup = new EpollEventLoopGroup(nettyConfig.nettyServerBossThreadCount);
+        workerGroup = new EpollEventLoopGroup(nettyConfig.nettyServerWorkerThreadCount);
+      } else {
+        bossGroup = new NioEventLoopGroup(nettyConfig.nettyServerBossThreadCount);
+        workerGroup = new NioEventLoopGroup(nettyConfig.nettyServerWorkerThreadCount);
+      }
```
Comment (justinlin-linkedin) [inline ambry-rest/src/main/java/com/github/ambry/rest/NettyServer.java:84]: while we are changing the EventLoopGroup to Epoll, we should probably change the ServerSocketChannel to Epoll as well. I suspect that crash was caused by the incompatible EventLoopGroup with ServerSocketChannel.

Comment (justinlin-linkedin) [reply ambry-rest/src/main/java/com/github/ambry/rest/NettyServer.java:84]: https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/nio/AbstractNioChannel.java#L372 This should be where the check is performed.

Comment (gaojieliu) [reply ambry-rest/src/main/java/com/github/ambry/rest/NettyServer.java:84]: That is right, the epoll usage should be consistent.

**Comment by cgtz** [comment]
Hi @zzmao,

I saw some test failures in the most recent travis-ci build in ServerHttp2Test:
```
com.github.ambry.server.ServerHttp2Test > endToEndTest[0] FAILED
    java.lang.AssertionError: 
        at org.junit.Assert.fail(Assert.java:91)
        at org.junit.Assert.fail(Assert.java:98)
        at com.github.ambry.server.ServerTestUtil.endToEndTest(ServerTestUtil.java:561)
        at com.github.ambry.server.ServerHttp2Test.endToEndTest(ServerHttp2Test.java:97)
com.github.ambry.server.ServerHttp2Test > endToEndTest[1] STARTED
com.github.ambry.server.ServerHttp2Test > endToEndTest[1] FAILED
    java.lang.AssertionError: 
        at org.junit.Assert.fail(Assert.java:91)
        at org.junit.Assert.fail(Assert.java:98)
        at com.github.ambry.server.ServerTestUtil.endToEndTest(ServerTestUtil.java:561)
        at com.github.ambry.server.ServerHttp2Test.endToEndTest(ServerHttp2Test.java:97)
```
Here is the build scan: https://gradle.com/s/2gdtl3ehuizcy

I restarted the test to see if it fails again

**Comment by cgtz** [comment]
It will probably fail again since this was the issue:
```
java.lang.IllegalStateException: incompatible event loop type: io.netty.channel.nio.NioEventLoop
	at io.netty.channel.AbstractChannel$AbstractUnsafe.register(AbstractChannel.java:462)
	at io.netty.channel.SingleThreadEventLoop.register(SingleThreadEventLoop.java:87)
	at io.netty.channel.SingleThreadEventLoop.register(SingleThreadEventLoop.java:81)
	at io.netty.channel.MultithreadEventLoopGroup.register(MultithreadEventLoopGroup.java:86)
	at io.netty.bootstrap.AbstractBootstrap.initAndRegister(AbstractBootstrap.java:311)
	at io.netty.bootstrap.Bootstrap.doResolveAndConnect(Bootstrap.java:157)
	at io.netty.bootstrap.Bootstrap.connect(Bootstrap.java:118)
	at com.github.ambry.network.http2.Http2BlockingChannel.connect(Http2BlockingChannel.java:113)
	at com.github.ambry.server.ServerTestUtil.endToEndTest(ServerTestUtil.java:187)
	at com.github.ambry.server.ServerHttp2Test.endToEndTest(ServerHttp2Test.java:97)
```
https://scans.gradle.com/s/2gdtl3ehuizcy/tests/6awkxu3uswmr6-eskuugjgq7tug

**Pull request merged**